extensions/default/src/DicomWebDataSource/index.js

```diff
@@ -24,123 +22,129 @@ const { naturalizeDataset, denaturalizeDataset } = DicomMetaDictionary;
-          dicomWebConfig.omitQuotationForMultipartRequest
-        );
-
-        return {
-          ...authorizationHeader,
-          Accept: formattedAcceptHeader,
-        };
-      };
-
-      qidoConfig = {
-        url: dicomWebConfig.qidoRoot,
-        staticWado: dicomWebConfig.staticWado,
-        singlepart: dicomWebConfig.singlepart,
-        headers: userAuthenticationService.getAuthorizationHeader(),
-        errorInterceptor: errorHandler.getHTTPErrorHandler(),
-      };
-
-      wadoConfig = {
-        url: dicomWebConfig.wadoRoot,
-        staticWado: dicomWebConfig.staticWado,
-        singlepart: dicomWebConfig.singlepart,
-        headers: userAuthenticationService.getAuthorizationHeader(),
-        errorInterceptor: errorHandler.getHTTPErrorHandler(),
-      };
-
-      // TODO -> Two clients sucks, but its better than 1000.
-      // TODO -> We'll need to merge auth later.
-      qidoDicomWebClient = dicomWebConfig.staticWado
-        ? new StaticWadoClient(qidoConfig)
-        : new api.DICOMwebClient(qidoConfig);
-
-      wadoDicomWebClient = dicomWebConfig.staticWado
-        ? new StaticWadoClient(wadoConfig)
-        : new api.DICOMwebClient(wadoConfig);
+      clientManager = new ClientManager({
+        params,
+        query,
+        dicomWebConfig,
+        userAuthenticationService,
+      });
+      // here clientManager returns, if any, the reject function of the first
+      // client that supports it. This server should be the first main server
+      implementation.clientCanReject = clientManager.clientCanReject();
     },
     query: {
       studies: {
         mapParams: mapParams.bind(),
         search: async function (origParams) {
-          qidoDicomWebClient.headers = getAuthrorizationHeader();
-          const { studyInstanceUid, seriesInstanceUid, ...mappedParams } =
-            mapParams(origParams, {
-              supportsFuzzyMatching: dicomWebConfig.supportsFuzzyMatching,
-              supportsWildcard: dicomWebConfig.supportsWildcard,
-            }) || {};
-
-          const results = await qidoSearch(qidoDicomWebClient, undefined, undefined, mappedParams);
-
+          clientManager.setQidoHeaders();
+          let results = [];
+          const studyInstanceUIDs = {};
+          // concatenate series metadata from all servers
+          const clients = clientManager.getClients();
+          for (let i = 0; i < clients.length; i++) {
+            const { studyInstanceUid, seriesInstanceUid, ...mappedParams } =
+              mapParams(origParams, {
+                supportsFuzzyMatching: clients[i].supportsFuzzyMatching,
+                supportsWildcard: clients[i].supportsWildcard,
+              }) || {};
+            let clientResults;
+            try {
+              clientResults = await qidoSearch(
+                clients[i].qidoDicomWebClient,
+                undefined,
+                undefined,
+                mappedParams
+              );
```

Comment:
This is not a good pattern for looped async calls. You should .map and promise.all after to let the requests sent in parallel